src/charm.py

```diff
@@ -578,13 +580,21 @@ def _alerting_config(self) -> str:
 
         return ",".join(alertmanagers)
 
+    @property
     def _get_v12_migration_date(self) -> str:
-        """Get the 'from' date from the v12 schema in Loki config."""
+        """Get the 'from' date from the v12 schema in Loki config.
+
+        In 2024-05 we changed the storage backend the charm configures from v11/boltdb to v12/tsdb.
+
+        Returns:
+        str: If v12 schema is found, returns the 'from' date of the v12 schema
+                        in ISO 8601 format (YYYY-MM-DD), otherwise returns empty string.
+        """
         if not self._loki_container.can_connect():
             return ""
         try:
             running_config = yaml.safe_load(
-                self._loki_container.pull(LOKI_CONFIG_PERSISTED, encoding="utf-8").read()
+                self._loki_container.pull(LOKI_CONFIG_BACKUP, encoding="utf-8").read()
             )
         except PathError:
```

Comment:
We may end up needing more exception types here: sometimes we get pebble errors after a can_connect guard. I think ConnectionError, APIError and maybe another one.

But a yaml parsing error we probably want to have the charm in error state?

Reply:
so most of these are already taken care of within the [can_connect](https://github.com/canonical/operator/blob/6cbc264a3bba177a34d93e8422e7f9c0807a0961/ops/model.py#L2132) guard, the parsing error doesn't seem recoverable tho so it makes sense to have it in error state

Reply:
That is true but there is a subtle difference: it is only the `can_connect` call that is inside the try-except - the code that follows inside your if block is not. `can_connect` is a point-in-time check, i.e. if you call it twice, the second may return false. So the code after a can_connect guard can still fail for pebble operations. In in our experience it does (transient startup error; mentioned [here](https://discourse.charmhub.io/t/its-probably-ok-for-a-unit-to-go-into-error-state/13022)). 

Based on the current semantics of this function, it seems you'd want to return `""` on pebble.ConnectionError, etc.

Zooming out a bit, I wonder if returning `""` on error is the best way to approach this.